nototools/report_coverage_data.py

```diff
@@ -0,0 +1,286 @@
+  cp_limit = 1000
+  chart_width = 250
+  chart_row_height = 8
+
+  HEADER = """<!DOCTYPE html>
+<html lang="en">
+<head>
+  <meta charset="utf-8">
+  <title>Block coverage</title>
+  <style>
+  td.range {font-family: monospace; text-align: right;}
+  td.count {font-family: monospace; text-align: right;}
+  td.pct {font-family: monospace; text-align: right;}
+  div.chart {width:%dpx; position:relative;}
+  div.tot {border: 1px solid black; padding: 1px;}
+  div.bar {height: %dpx; background-color: green;}
+  </style>
+</head>
+<body>
+  <h2>Block coverage</h2>
+""" % (chart_width, chart_row_height)
+  FOOTER = """</body>
+</html>
+"""
+
+  out_file.write(HEADER)
+  if msg:
+    out_file.write('  <p>')
+    out_file.write(msg)
+    out_file.write('\n')
+  block_data.sort()
+  out_file.write(
+      '  <table>\n'
+      '    <tr><th colspan=3>&nbsp;')
+  for name in names:
+    out_file.write('<th colspan=3>%s' % name)
+  out_file.write('\n')
+  out_file.write(
+      '    <tr><th>Range<th>Block Name<th>Count')
+  for _ in names:
+    out_file.write('<th>Count<th>Pct<th>Chart')
+  out_file.write('\n')
+
+  max_block_size = max(len(t[3]) for t in block_data)
+  if max_block_size > cp_limit:
+    max_block_size = cp_limit
+  for start, end, name, block_cps, block_covered_cps_list in block_data:
+    range_str = '%04x-%04x' % (start, end)
+    num_in_block = len(block_cps)
+    tot_ht = 1 + int(math.floor(float(num_in_block) / cp_limit))
+    tot_px = int(float(chart_width) * num_in_block / max_block_size / tot_ht)
+    vir_cp_limit = int(float(num_in_block) / tot_ht)
+    out_file.write('    <tr>')
+    out_file.write('<td class="range">')
+    out_file.write(range_str)
+    out_file.write('<td class="name">')
+    out_file.write(name)
+    out_file.write('<td class="count">')
+    out_file.write(str(num_in_block))
+    for block_covered_cps in block_covered_cps_list:
+      num_covered = len(block_covered_cps)
+      pct = '%d%%' % int(100.0 * num_covered / num_in_block)
+      bar_ht = min(
+          tot_ht, 1 + int(math.floor(float(num_covered) / vir_cp_limit)))
+      bar_px = int(float(chart_width) * num_covered / max_block_size / bar_ht)
+      out_file.write('<td class="count">')
+      out_file.write(str(num_covered))
+      out_file.write('<td class="pct">')
+      out_file.write(pct)
+      out_file.write('<td><div class="chart">')
+      out_file.write('<div class="tot" style="width:%dpx; height:%dpx">' % (
+          tot_px, tot_ht * chart_row_height))
+      out_file.write('<div class="bar" style="width:%dpx; height:%dpx">' % (
+          bar_px, bar_ht * chart_row_height))
+      out_file.write('</div></div></div>\n')
+  out_file.write(FOOTER)
+
+
+def write_block_coverage_text(block_data, names, msg, out_file=sys.stdout):
+  block_data.sort()
+  print >> out_file, msg
+  name_len = max(len(t[2]) for t in block_data)
+  fmt_str = '%%%ds' % name_len
+  fmt_str = '%13s ' + fmt_str + ' %5s'
```

Comment:
I could, I guess.  I'd have to parenthesize the first line inside the second expression or else I think the string concatenation would take precedence.  I don't think python supports formatting with field-width arguments like C printf, so it's a bit messy no matter what I do.  I'll just leave it.